packages/react-async/src/context/assets.ts

```diff
@@ -1,23 +1,71 @@
 import * as React from 'react';
 import {EffectKind} from '@shopify/react-effect';
+import {AssetTiming} from '../types';
+
+export interface AssetSelector {
+  id: string;
+  styles: boolean;
+  scripts: boolean;
+}
+
+interface AssetOptions {
+  styles: AssetTiming;
+  scripts: AssetTiming;
+}
 
 export const EFFECT_ID = Symbol('react-async');
 
 export class AsyncAssetManager {
-  readonly used = new Set<string>();
-
   readonly effect: EffectKind = {
     id: EFFECT_ID,
     betweenEachPass: () => {
-      this.used.clear();
+      this.assets.clear();
     },
   };
 
-  markAsUsed(id: string) {
-    this.used.add(id);
+  private assets = new Map<string, AssetOptions>();
+
+  used(timing: AssetTiming | AssetTiming[] = AssetTiming.Immediate) {
+    const timingArray = Array.isArray(timing) ? timing : [timing];
+
+    const assets: AssetSelector[] = [];
+
+    for (const [asset, {scripts, styles}] of this.assets) {
+      const scriptsMatch = timingArray.includes(scripts);
+      const stylesMatch = timingArray.includes(styles);
+
+      if (stylesMatch || scriptsMatch) {
+        assets.push({id: asset, scripts: scriptsMatch, styles: stylesMatch});
+      }
+    }
+
+    return assets;
+  }
+
+  markAsUsed(
+    id: string,
+    timing:
+      | AssetTiming
+      | {scripts?: AssetTiming; styles?: AssetTiming} = AssetTiming.Immediate,
+  ) {
+    const current = this.assets.get(id);
+    const scripts = typeof timing === 'object' ? timing.scripts : timing;
+    const styles = typeof timing === 'object' ? timing.styles : timing;
+
+    if (current == null) {
+      this.assets.set(id, {
+        scripts: scripts || AssetTiming.Immediate,
+        styles: styles || AssetTiming.Immediate,
+      });
+    } else {
+      this.assets.set(id, {
+        scripts: Math.max(scripts || current.scripts, current.styles),
```

Comment:
This max/min stuff is made me go reread a bunch because I suddenly felt like I had no idea what this is doing.

Might be a good spot for a 'why' comment.

Reply:
Agreed :+1: